docs/doxygen/man/man3/GetMessageList.3
.TH "GetMessageList" 3 "Thu Aug 30 2018" "esc" \" -*- nroff -*-
.ad l
.nh
.SH NAME
GetMessageList \- Get message list command class\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <getmessagelist\&.h>\fP
.PP
Inherits \fBBlockCommand\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBGetMessageList\fP (uint16_t abank, uint32_t auser, uint32_t block, uint32_t time)"
.br
.ti -1c
.RI "virtual int \fBgetType\fP () override"
.br
.RI "\fIReturn TXSTYPE_MSG as type \&. \fP"
.ti -1c
.RI "virtual CommandType \fBgetCommandType\fP () override"
.br
.RI "\fIReturn eReadingOnly as command type \&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetData\fP () override"
.br
.RI "\fIGet pointer to command data structure\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetResponse\fP () override"
.br
.RI "\fIGet pointer to response data\&. \fP"
.ti -1c
.RI "virtual void \fBsetData\fP (char *data) override"
.br
.RI "\fIPut data as a char list and put convert it to data structure\&. \fP"
.ti -1c
.RI "virtual void \fBsetResponse\fP (char *response) override"
.br
.RI "\fIApply data as a response struct\&. \fP"
.ti -1c
.RI "virtual int \fBgetDataSize\fP () override"
.br
.RI "\fIGet data struct size\&. Without signature\&. \fP"
.ti -1c
.RI "virtual int \fBgetResponseSize\fP () override"
.br
.RI "\fIGet response data struct size\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetSignature\fP () override"
.br
.RI "\fIGet pointer to signature data\&. \fP"
.ti -1c
.RI "virtual int \fBgetSignatureSize\fP () override"
.br
.RI "\fIGet signature size\&. \fP"
.ti -1c
.RI "virtual void \fBsign\fP (const uint8_t *hash, const uint8_t *sk, const uint8_t *pk) override"
.br
.RI "\fISign actual data plus hash using user private and public keys\&. \fP"
.ti -1c
.RI "virtual bool \fBcheckSignature\fP (const uint8_t *hash, const uint8_t *pk) override"
.br
.RI "\fICheck actual signature\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetTime\fP () override"
.br
.RI "\fIGet time of command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserId\fP () override"
.br
.RI "\fIGet User ID\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetBankId\fP () override"
.br
.RI "\fIGet Node ID\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetFee\fP () override"
.br
.RI "\fIGet command fee\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetDeduct\fP () override"
.br
.RI "\fIGet change in cash balance after command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserMessageId\fP () override"
.br
.RI "\fIGet message id\&. \fP"
.ti -1c
.RI "virtual bool \fBsend\fP (\fBINetworkClient\fP &netClient) override"
.br
.RI "\fISend data to the server\&. \fP"
.ti -1c
.RI "virtual void \fBsaveResponse\fP (\fBsettings\fP &sts) override"
.br
.RI "\fISave command response to settings object\&. \fP"
.ti -1c
.RI "virtual std::string \fBtoString\fP (bool pretty) override"
.br
.ti -1c
.RI "virtual void \fBtoJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual void \fBtxnToJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual std::string \fBusageHelperToString\fP () override"
.br
.ti -1c
.RI "uint32_t \fBgetBlockTime\fP ()"
.br
.ti -1c
.RI "virtual \fBErrorCodes::Code\fP \fBprepareResponse\fP ()"
.br
.RI "\fIGet data from msglist file\&. \fP"
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "\fBGetMessageListData\fP \fBm_data\fP"
.br
.ti -1c
.RI "uint8_t \fBm_responseTxnHash\fP [SHA256_DIGEST_LENGTH]"
.br
.ti -1c
.RI "std::vector< \fBMessageRecord\fP > \fBm_responseMessageList\fP"
.br
.ti -1c
.RI "\fBcommandresponse\fP \fBm_response\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Get message list command class\&. 
.PP
Definition at line 10 of file getmessagelist\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "bool GetMessageList::checkSignature (const uint8_t * hash, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Check actual signature\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 64 of file getmessagelist\&.cpp\&.
.SS "bool GetMessageList::send (\fBINetworkClient\fP & netClient)\fC [override]\fP, \fC [virtual]\fP"

.PP
Send data to the server\&. 
.PP
\fBParameters:\fP
.RS 4
\fInetClient\fP Netwrok client implementation of \fBINetworkClient\fP interface\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 95 of file getmessagelist\&.cpp\&.
.SS "void GetMessageList::sign (const uint8_t * hash, const uint8_t * sk, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Sign actual data plus hash using user private and public keys\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIsk\fP Pointer to provate key\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 60 of file getmessagelist\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for esc from the source code\&.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2018-08-24
### Changed
- Genesis file for the launch
- Update README

## [0.0.6] - 2018-08-10
### Added
- Profit sharing among TOP and VIP nodes

### Changed
- Genesis file for test launch

### Fixed
- Vote counting in get_block command
- Ability to send externally signed transactions

## [0.0.5] - 2018-08-01
### Added
- Support passing version by cmake command.
- Contributing guide
- "Dry run" test case

### Fixed
- Node stops after incorrect `send_many` request  - #134
- Fee checking
- Internal transactions numbers

## [0.0.4] - 2018-07-27
### Changed
- Update default parameters
- Creating a dev version with reference to the last tag

[Unreleased]: https://github.com/adshares/ads/compare/v1.0.0...HEAD

[1.0.0]: https://github.com/adshares/ads/compare/v0.0.6...v1.0.0
[0.0.6]: https://github.com/adshares/ads/compare/v0.0.5...v0.0.6
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/adshares/ads/compare/v0.0.3...v0.0.4


src/common/command/getmessagelist.cpp
#include "getmessagelist.h"

#include "ed25519/ed25519.h"
#include "abstraction/interfaces.h"
#include "helper/json.h"
#include "helper/hash.h"
#include "hash.hpp"
#include "parser/msglistparser.h"

GetMessageList::GetMessageList()
    : m_data{} {
    m_responseError = ErrorCodes::Code::eNone;
}

GetMessageList::GetMessageList(uint16_t abank, uint32_t auser, uint32_t block, uint32_t time)
    : m_data(abank, auser, block, time) {
    m_responseError = ErrorCodes::Code::eNone;
}

unsigned char* GetMessageList::getData() {
    return reinterpret_cast<unsigned char*>(&m_data.info);
}

unsigned char* GetMessageList::getResponse() {
    return reinterpret_cast<unsigned char*>(&m_response);
}

void GetMessageList::setData(char* data) {
    m_data = *reinterpret_cast<decltype(m_data)*>(data);
}

void GetMessageList::setResponse(char* response) {
    m_response = *reinterpret_cast<decltype(m_response)*>(response);
}

int GetMessageList::getDataSize() {
    return sizeof(m_data.info);
}

int GetMessageList::getResponseSize() {
    return sizeof(m_response);
}

unsigned char* GetMessageList::getSignature() {
    return m_data.sign;
}

int GetMessageList::getSignatureSize() {
    return sizeof(m_data.sign);
}

int GetMessageList::getType() {
    return TXSTYPE_MGS;
}

CommandType GetMessageList::getCommandType() {
    return CommandType::eReadingOnly;
}

void GetMessageList::sign(const uint8_t* /*hash*/, const uint8_t* sk, const uint8_t* pk) {
    ed25519_sign(getData(), getDataSize(), sk, pk, getSignature());
}

bool GetMessageList::checkSignature(const uint8_t* /*hash*/, const uint8_t* pk) {
    return (ed25519_sign_open(getData(), getDataSize(), pk, getSignature()) == 0);
}

void GetMessageList::saveResponse(settings& /*sts*/) {
}

uint32_t GetMessageList::getUserId() {
    return m_data.info.src_user;
}

uint32_t GetMessageList::getBankId() {
    return m_data.info.src_node;
}

uint32_t GetMessageList::getTime() {
    return m_data.info.ttime;
}

int64_t GetMessageList::getFee() {
    return 0;
}

int64_t GetMessageList::getDeduct() {
    return 0;
}

uint32_t GetMessageList::getUserMessageId() {
    return 0;
}

bool GetMessageList::send(INetworkClient& netClient) {
    if(!netClient.sendData(getData(), sizeof(m_data))) {
        ELOG("GetMessageList sending error\n");
        return false;
    }

    if (!netClient.readData((int32_t*)&m_responseError, ERROR_CODE_LENGTH)) {
        ELOG("GetMessageList reading error\n");
        return false;
    }

    if (m_responseError) {
        return true;
    }

    uint32_t no_of_msg = 0;
    if (!netClient.readData((int32_t*)&no_of_msg, sizeof(no_of_msg))) {
        ELOG("GetMessageList reading no of messages error\n");
        return false;
    }

    if (!netClient.readData((uint8_t*)&m_responseTxnHash, sizeof(m_responseTxnHash))) {
        ELOG("GetMessageList reading txn has error\n");
        return false;
    }

    for (unsigned int i=0; i<no_of_msg; ++i) {
        MessageRecord record;
        if (!netClient.readData((char*)&record, sizeof(record))){
            ELOG("GetMessageList reading %d record error\n", i);
            return false;
        }
        m_responseMessageList.push_back(record);
    }

    return true;
}

uint32_t GetMessageList::getBlockTime() {
    return m_data.info.block;
}

ErrorCodes::Code GetMessageList::prepareResponse() {
    Parser::MsglistParser msglistParser(m_data.info.block);
    if (!msglistParser.load()) {
        return ErrorCodes::Code::eNoMessageListFile;
    } else {
        std::copy(msglistParser.getHeaderHash(),
                  msglistParser.getHeaderHash()+sizeof(m_responseTxnHash),
                  m_responseTxnHash);
        m_responseMessageList = msglistParser.getMessageList();
    }
    return ErrorCodes::Code::eNone;
}

std::string GetMessageList::toString(bool /*pretty*/) {
    return "";
}

void GetMessageList::toJson(boost::property_tree::ptree& ptree) {
    if (m_responseError) {
        ptree.put(ERROR_TAG, ErrorCodes().getErrorMsg(m_responseError));
    } else {
        char blockhex[9];
        blockhex[8]='\0';
        sprintf(blockhex,"%08X", m_data.info.block);
        ptree.put("block_time_hex", blockhex);
        ptree.put("block_time", m_data.info.block);

        char hash[65];
        hash[64]='\0';
        ed25519_key2text(hash, m_responseTxnHash,32);
        ptree.put("msghash",hash);
        ptree.put("message_count",m_responseMessageList.size());
        hashtree htree;
        boost::property_tree::ptree messages;
        for (auto &it : m_responseMessageList) {
            boost::property_tree::ptree msghash;
            ed25519_key2text(hash, it.hash,32);
            msghash.put_value(Helper::print_msg_pack_id(it.node_id, it.node_msid));
            htree.update(it.hash);
            messages.push_back(std::make_pair("",msghash));
        }
        uint8_t fullHash[SHA256_DIGEST_LENGTH];
        htree.finish(fullHash);


        if(memcmp(m_responseTxnHash, fullHash, SHA256_DIGEST_LENGTH)) {
            ed25519_key2text(hash, fullHash, SHA256_DIGEST_LENGTH);
            ptree.put("msghash_calculated",hash);
            ptree.put("confirmed","no");
        } else {
            ptree.put("confirmed","yes");
        }
        if(m_responseMessageList.size() > 0) {
            ptree.add_child("messages",messages);
        }
    }
}

void GetMessageList::txnToJson(boost::property_tree::ptree& ptree) {
    using namespace Helper;
    ptree.put(TAG::TYPE, getTxnName(m_data.info.ttype));
    ptree.put(TAG::SRC_NODE, m_data.info.src_node);
    ptree.put(TAG::SRC_USER, m_data.info.src_user);
    ptree.put(TAG::TIME, m_data.info.ttime);
    ptree.put(TAG::BLOCK, m_data.info.block);
    ptree.put(TAG::SIGN, ed25519_key2text(getSignature(), getSignatureSize()));
}

std::string GetMessageList::usageHelperToString() {
    std::stringstream ss{};
    ss << "Usage: " << "{\"run\":\"get_message_list\",[\"block\":<block time as hex>]}" << "\n";
    ss << "Example: " << "{\"run\":\"get_message_list\",\"block\":\"5B052CE0\"}" << "\n";
    return ss.str();
}
